### ISSUE TITLE
wgsl: language evolution: managing rollout of core (essentially sugar) language features

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#3149]
This is forked from #2941

In short, Google's proposal is at: https://github.com/gpuweb/gpuweb/issues/2941#issuecomment-1160877990 
- classifying features into sugar vs. non-sugar.  Sugar features can be implemented anywhere. (E.g 'do-while' loops).
- bundle sugar features in to monotonically increasing shells of functionality
- use Semver conventions of backward compatibility for those increasing shells of functionality

Further discussion 

* 2022-06-01 meeting: https://github.com/gpuweb/gpuweb/issues//2941#issuecomment-1164897200
* @kdashg 's proposal for std namespacing and automatically importing them, including a scheme for increasing shells of functionality: https://github.com/gpuweb/gpuweb/issues//2941#issuecomment-1164927025
* 2022-06-28 meeting: https://github.com/gpuweb/gpuweb/issues//2941#issuecomment-1170249066

Temporarily putting this in v1.0 milestone, for triage.   I'd like an affirmative discussion about milestone.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-07-26-minutes#wgsl-language-evolution-managing-rollout-of-core-essentially-sugar-language-features-3149">WGSL meeting minutes 2022-07-26</a></summary>

* KG: Want MM for this, so we table. 

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-09-minutes#wgsl-language-evolution-managing-rollout-of-core-essentially-sugar-language-features-3149">WGSL 2022-08-09 Minutes</a></summary>

* KG: Rollup with links to other discussions. In V1 so will get triaged. Sounds like in short need solution to 1) do we need to figure out for v1? Seems like we should, but depends on if we feel like we need to do something different or if we can keep doing what we’re doing
* MM: Urgent question is if user defined functions shadow standard library, if answer is yes then don’t need to decide now
* AB: We agreed on shadowing predeclared builtins. Think important part is compat going forward and settling on maintaining backwards compat. If we agree on that then don’t have to worry about it.
* MM: If there is a breaking change it has to be opt-in, standard on web platform.
* KG: That’s what made me satisfied, we’d be using opt-in for breaking, for non-breaking weight on scale and decide if they need an indication. User defined functions shadowing makes easier. 
* DN: And also shadow predeclared types like i32 and f32. In doing that and removing keywords, couple PRs posted and a technical problem comes up where mat2x2&lt;f32> how do you resolve as a type constructor or function call. Need to invent words. Types as well is main point.
* JB: Maybe in the future we define generics which might shadow wgsl types
* MM: Could imagine that
* DN: Need name for thing with type parameterization which needs a type.
* MM: The stdlib can have overload, so if I make a foo and stdlib has a foo, do i shadow all of them?
* DN: Yes
* AB: Yes.
* KG: Today yes. Answer is probably to allow overload which we do in the stdlib anyway. How much work is user defined overloads?
* JB: Don’t think that’s the problem, question is if you get confusion as some calls go to user function and some to stdlib. Would be confusing. If there is a redefinition then that shadows all overrides or things from other scopes. Mixing depending on types of arguments would be bad.
* MM: Point of shadowing is the author makes function and calls and won’t accidentally call stdlib. If we don’t fully shadow then adding to stdlib might take over in precedence rules.
* KG: Do we have precedence resolution?
* MM: We don’t because the only overloads are in the stdlib and don’t need general system, if decide users can do it then will need a general system.
* KG: If decide to never support, like said functions never coerce their types. Then overload resolution is find the one for the type.
* DS: What about abstract?
* KG: Can you name it
* AB: Through const
* MM: Imagine sample function taking offset int and offset float. That’s the scary part.
* KG: Don’t want that anyway?
* **TODO: Shadowing overloaded builtins?**

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-16-minutes#wgsl-language-evolution-managing-rollout-of-core-essentially-sugar-language-features-3149">WGSL 2022-08-16 Minutes</a></summary>

* Previously: TODO: Shadowing overloaded builtins?
* MM: I thought we had a resolution on this last week. Are we not done here?
* KG: If we did not in minutes
* MM: Previous resolution, **because user types and user functions can shadow builtins we don’t have to solve now we can solve later**
* KG: Even for overloaded
* MM: Yes. **Any user function shadows all overload builtins**.
* KG: That’s fine. Didn’t make it into the minutes, so didn’t see it. **Resolved**.

</details>

**Comment by kdashg** [comment]
The part that we decided we need was done in #2941, and the rest is now post-v1.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-12-06-minutes#language-evolution-httpsgithubcomgpuwebgpuwebissues3149-">WGSL 2022-12-06 Minutes</a></summary>

* Technically we *can* release v1 without it but it sure would be nice to have a plan.
* Eg. a min-version-sensing function for use in a static assert.
* DN: We stopped talking about this for a while. We can ship v1 without a plan, but it would be unfortunate if we need to add something to check the version, but then it’s not available in the first version, so even when you’re checking the version you need some dancing to get around its absence in v1. **Google is working on this internally.**
* AB: Myles had talking points about this, so we can’t reach any conclusion about this until he returns, but **we can work on it**.

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-12-13-minutes#language-evolution-httpsgithubcomgpuwebgpuwebissues3149-">WGSL 2022-12-13 Minutes</a></summary>

* AB: Happy restate what we think is the direction.
    * We (W3C WGSL)  increments a number for core version, and say what feature is in each version.
* AB: Know Myles was against that. What do other stakeholders think?
* JB: Idea of letting developers specify an intended high water mark and have the browser flag if they’ve gone beyond the intended limit.  Sounds cool. As long as we’re not incrementing major version frequently. (Those should be very rare). But as long as the minor version bumps have the preserving properties you’re looking for. It wasn’t clear to us how prototyping should work.  Give to users something to prototype but are not in the spec yet. How do those fit in the picture?  Also seems to run afoul of what Apple said they wanted: devs try stuff and feature detect by seeing pass/fail.  Perhaps wait for Myles to return.
* AB: Agree, not trying to get a decision before Myles return.
* AB: Experimental features are not mentioned in the spec.  Expect they would work like other we specs:  not documented in the spec, and features have browser-specific previx/suffix. And if those don’t land eventually in the spec, then that’s your users’s problems.  We want to reduce the granularity of the minor number; expect browsers to implement the incremental features without much phase difference. (Shouldn’t have one feature take a month and another take 5 years.)  Won’t have too many “this is on but that is off” interactions in the features. That’s why we think the bundling is a reasonable approach.
* JB: We like the way it avoids the combinatorial explosion. It’s just linear.  We’re not too concerned about forcing a linear order.
* AB: If linear ordering becomes a problem then that’s a sign we’re picking the wrong things to go into core.

</details>

**Comment by litherum** [comment]
Our biggest concern is having to implement and maintain `n` different compilers inside our implementation. We would object to GLSL's model, where the author tells the browser which version to use, because it violates this concern. (`"use strict"` would also be an example of this antipattern.)

For hardware-gated features, extensions or `enable` lines are, of course, a requirement.

For non-hardware-gated features (called "sugar" elsewhere), we prefer to mirror the JavaScript (ECMAScript) model for language evolution (which also coincidentally more-or-less matches the CSS model, too). Specifically, one where the author doesn't tell the browser anything about versioning information, and the browser does its best to honor the sources it was given. Feature detection is possible.

There are 2 reasons for this preference:
1. There are simply fewer complexities for authors, in general. They don't have to worry about supplying extra metadata along with their shader (either in-band or out-of-band), which have to be maintained and kept consistent with the sources.
2. If the author gets the version number wrong, but still write a valid program, it would be a disservice for the browser to reject the program, that it can fully understand, just because the author didn't do their bookkeeping correctly.

For background info, JavaScript authors (and CSS authors) do one of 3 things:
1. Look at caniuse or other sites, determine that some feature is supported in all their target browsers, and then just use it in their sources unconditionally. No version numbers required
2. Use `eval()` or `@supports` to do feature detection for a particular feature. (For WGSL, this would take the form of `createShaderModule()`.)
3. Use an offline transpiler like [Babel](https://babeljs.io) (or [Sass](https://sass-lang.com), to some extent) to translate new sources into old sources.

Javascript (really, ECMAScript) does have [versions](https://en.wikipedia.org/wiki/ECMAScript_version_history), but they are not exposed anywhere. Instead, browsers simply just try to implement all the features in a single version at roughly the same time. However, this is unenforced; it's totally possible for a browser to ship while supporting just half of the latest version of the language. This is generally a good thing; it means unrelated features don't hold back shipment of other features.

**Comment by mehmetoguzderin** [comment]
I'd also highly support JavaScript and CSS models. If the language opts into such evolution, are there any lessons that we can learn from languages with such a roll-out & fortify the current approach to V1 to ensure a flexible baseline for the future where people won't need to come into the tricky or stressful dialogue of challenging something that came bundled with the inception (anything blocks incrementality of the language by any axis)? It is a fact that syntactic ambiguity tests probably prepare a bit (thanks to the efforts of @dneto0 to realize the analyzer), though even there, we might have some things to enable incremental development (such as https://github.com/gpuweb/gpuweb/issues/3520 to prevent changes to baseline grammar rules).

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-01-10-minutes#language-evolution-httpsgithubcomgpuwebgpuwebissues3149-">WGSL 2023-01-10 Minutes</a></summary>

* AB: I recall, a while back Myles had queued up discussion points around this issue.  Would like to hear them if available. Google has expressed an opinion (e.g. high water mark of used features).
* MM: Don’t have queued up issues. Did post to the issue last night.
* AB: When we’ve discussed the JavaScript model internally. On the API side, querying when things exist or not makes sense. On the shader side it doesn’t seem great to do a bunch of compile-fail.  We agree with overall major point: don’t want version hell.
* MM: True the “version hell” is our primary concern.   The larger model is concern #2.
* AB: Definitely agree with the major concern.
* KG: Hard to tell lessons learned from WebGL, WebGL 2. While we do have version declarations in the GLSL that are not strictly compatible; but perhaps it’s not too onerous?
* MM: Can speak more why about the primary concern. First, it’s way easier to maintain compiler if you don’t have to maintain all the version layers.  (So benefit to us, to reduce complexity).  For developer learning, searching for tutorial for WebGL, many of them are very old.
* KG: Does that not run counter to the desire that whatever is on tip is the “real one”. Isn’t it a benefit to see “ah, this example is old”.
* MM: E.g. look for GLSL tutorial that may say “varying”. Would like the JS model where wouldn’t break backward compatibility.
* DN: Isn’t that concern about breaking backward compatibility?
* AB: Really the concern is when introducing new things that are close to old functionality. Don’t want 50 ways to do very similar things.
* MM: Can see “do we have version numbers” is distinct from “changing keyword”. But I will just make the observation that GLSL started with version numbers, and then felt free to rename keywords.
* KG:  Can agree we should agree it’s bad to rename keywords. I’m not particularly concerned about slipping up: when we might do it, restate the argument, and it’s compelling.
* **Can think more, but nothing concrete to conclude in this meeting today.**

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-01-17-minutes#language-evolution-httpsgithubcomgpuwebgpuwebissues3149-">WGSL 2023-01-17 Minutes</a></summary>

* DN: Read comment from MM last week, thank you for stating clearly, haven't discussed internally due to other things
* MM: Defer or just in?
* KG: I think we're colessing here, I think some of the concerns some of us have are things we want to make sure are handled but aren't deal breaker problems.
* MM: Yea, also think the search space is pretty open and I bet there is some solution to keep everyone happy. Just haven't found that middle spot yet.
* DN: To  pull out one pice, don't want to maintain N versions for future sets. I agree with that, and my picture may not be correct, but there is a window of versions that the compiler may support and that window maybe as narrow as you like. Doesn’t have to go back to beginning of time. So, company A starts in 5 years time and only builds from that point. Think we want to allow narrow windows. Have not gone back to revisit. I agree you don't' want to, forever, maintain compatibility or the fact you'll reject old programs. It's ok to accept more then the min.
* MM: Assume narrow window is 4,5,6 and program comes along and says I"m v2. that's the interesting case.
* DN: Because requesting to not use anything after v2?
* MM: If asked for v2, what does that mean? What actions do we take? Would be a shame if the src identified itself as v2, and browser doesn't' support so reject.
* DS: Part of that is keeping backwards compat, a v6 compiler still compiles v2.
* MM: If we keep backwards compat indefinitely, why should the source even declare itself? Either you support or not?
* JB: Helps developers know when they overstepped their portability rules
* MM: Can be offline
* JB: It's a lint
* BC: What MM was saying was one of Googles original proposals. Add features and things keep compiling and we make sure things keep compiling. Don't want to support N different versions/feature sets. Don't want permutation hell. Having monotonic forward language and we agree what is in each monotonic increase is a nice way to proceed. The whole querying about version was folks concern about overstepping. Write program and accidentally use feature in newer version you didn't' expect, test browser and accidentally break things for a lagging browser. The whole I want to know what the highest water mark feature of my shader was to try to prevent that. Maybe linting would solve that. If there was an offline tool.
* JB: Moz hasn't talked about this, personally don't' have a strong feeling if this needs to be in browser or linter. Linter seems OK to me. (In response to why folks want this, wasn't saying the browser should do it).
* MM: Sounds like folks haven't thought very hard, maybe can't do full resolution, maybe can get consensus that feature dev of WGSL can be backwards compatible?
* KG: Goal is backwards compat. My feeling is eventually if we do feel it's valuable enough we cut a new version. But, going forward we're backwards compat.
* MM: Cool, benefit for a lot of folks. Thumbs up
* KG: That's the direction of consensus I think.
* JB: We know one thing we need is this user shadowing stuff that we've been working on. That's the only think I can think of right now. Are there other pieces of the spec that need spec language to make this rolling strategy work out? Or is just roll forward
* DS: That's the template thing
* AB: Thats a choice, we don't' have to have that feature
* MM: We should have the ability to have the feature.
* AB: We've said JS is the preprocessor, so we don't need to build future supports for it. Do we need template because we have JS? \
KG: Different shades on the gradient
* MM: Question on hand is, should we make a small change now so in the future we could have the opportunity for templates and generics
* KG: Overall on the same page.
* JB: Other spec work that's necessary?
* BC: Similar to the keywords, the enumerators on types is something we're having on types. They aren't reserved but how we resolve these is active discussion. If we have a parser generic thing that supports expressions and these enumerators do they incur shadowing as well?
* KG: What's an enumerator
* DN: Context dependant name `centroid` `rgba8unorm`. String used in specific spots, but isn't an identifier or keyword. If generalize we need to be able to use these declared, but not keyword name things.
* JB: We will not be able to add reserved words going forward. When are we making things work where we lean on reserved words?
* KG: Not to fill in the gaps too much it's important colour that we don't' have to positively never add keywords, there is the opportunity, like the array flattening stuff in JS ran into where they figured out a way forward where worried about splat. A question about balancing the forward compat of old websites.
* DS: Enables could enable new keywords
* BC: Getting into permutations of a naturally progressing language avoids, but that's one way to do it.
* (tabled)

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-01-24-minutes#language-evolution-httpsgithubcomgpuwebgpuwebissues3149-">WGSL 2023-01-24 Minutes</a></summary>

* KG: Main thing has to do with generics and going back to anything we want to talk about more on #3739.
* DN: Want to articulate one thing, we have made a dichotomy between an enable for hardware features and sugar is just implemented without opt-in. We also want to make the distinction that things go into core if we all think are valuable and we'd all implement. There is a middle ground where some think it's useful and others not, and having an enable for that is fine. Doesn't always have to be a hardware feature. 3 categories
    * Things in core, that hopefully everyone implements
    * Sugar landing at different times in different browsers
    * Hardware features
* MM: Little scary for the spec to say something that doesn't have consensus. Feel like the new category, would be explicitly non-standardized and the browser ships itself.
* KG: Like, you get it in spec even if not ready to implement. Like to have things in spec so we agree how it would work and address concerns and have conscience even if we're going to implement soon or not at all.
* AB: Think MM's concern is valid. Also wouldn't want spec to have a bunch of this is only for this platform. Littered with extensions which aren't widely available. If it's in the spec, maybe it's gated on hardware but expect it to be on all platforms. Can be experimental internal browser things to test. Things in some implementations that aren't in spec that have own enables, but side with MM that spec is widely available thing
* KG: Agree but caveat that would like to give feedback on what browsers ship and might accidentally cause to be built on top of the language. Browser might want to expand in a direction and browser gets enough mindshare that we want to ship, but another UA can't implement and we have a breaking change. Want to avoid cases like that. Understand not too many things one-off shipping. Experimentation happens outside spec. But don't encourage shipping things outside the spec
* AB: Agree to design what is going to ship. What goes in published WGSL spec vs not. Us having a design or PR is enough if everyone is commenting to not be in the spec.
* MM: Distinction I didn't describe properly, group collectively thinks it's a good idea and we want to move in a direction. Worried about someone coming up with a proposal, builds behind flag, and comes to group saying just for us, please put it into the WGSL spec
* KG: Happy to say no to that.
* MM: Also don't want to encourage folks to ship non-standard things, not what I meant.
* KG: Core is to sway judgment concerns. Fundamentally a judgment issue.
* MM: Maybe the right way to get conclusion everyone can agree with is that the 3rd category is ok as long as the group has consensus there should be an enable for the thing and that process is the same as all other processes.
* DN: Sounds right to me
* MM: Ties into pull request for vendor agnosticism and not land things in the spec that are prefixed.
* **MM: So are we good here for v1?**
* **(yes)**
* BC: **We need to determine a way to tell what version of WGSL supported**
* MM: Sounds like a query function and new issue. **Different from this issue**.
* BC: **Will discuss internally and may put up an api pull request**.
* KG: Suspect there will be a lot of appetite for a constant
* BC: What could you do with it? We don't have `if consteval` would need it in a block
* KG: Could const assert and error out.

</details>

**Comment by dneto0** [comment]
From discussion in the F2F 2023-02-16 ([raw notes](https://docs.google.com/document/d/1bbwoAu7NZ5GAWAE12xUNZi08fRuSRDfYLzHxV6yFDF8/edit#bookmark=id.mmb1492enidn))

* New features have names
* Add a 'require' directive that takes a list of feature names.
      * It captures the author intent that "This WGSL program deliberately uses these features".
      * It does not affect which features are enabled by the implementation.
* The API side exposes a list of supported features (as a list of strings)
      * This allows applications, and the CTS, to see what is supported.
      * Discussion: is this on the adaptor, or the device? Argument in favour of the adaptor is that these are "sugar" features.
* A "version" is modeled as a "feature" that is the union of other features, including any previous "version".

* Developer tools could add a linting option, which tells an author somehow if their shader uses a feature X, and their shader does not contain a 'require X;' directive.

**Comment by dneto0** [comment]
Re: "require" vs. "requires".

"enable" and "require" are ordinary verbs.
"enables" and "requires" are verbs in the 3rd person present tense.   

That's why I'm going with "require".

**Comment by ben-clayton** [comment]
A counter to `require`:
`enable` being an ordinary verb makes sense. It is a directive to the compiler to do something - enable a feature.

`require` is suggestive of something similar, like pulling in some additional logic, but that's not what it does. A shader `requires` features a, b, c, but it is not an directive for the compiler to do something, other than a pure assertion.

Offline you gave examples of perl and python using `require`, and suggested this is a reason for following suit - I consider that a good example of why it should not be called `require`, as this is entirely side effect free. It's essentially metadata.

**Comment by dneto0** [comment]
You've convinced me.

I also looked up Java's use of 'require'.

> 7.7.1 Dependencies

> The requires directive specifies the name of a module on which the current
module has a dependence.

So it's an assertion of a property, just like how we want to use it in WGSL.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-02-07-minutes#-discussions">WGSL 2023-02-07 Minutes</a></summary>

* [wgsl: language evolution: managing rollout of core (essentially sugar) language features #3149](https://github.com/gpuweb/gpuweb/issues/3149)
    * PR [#3792](https://github.com/gpuweb/gpuweb/pull/3792) Add wgsl_version_at_least builtin function
    * Discussion points:
        * Issue 0:  Get the *actual* version numbers instead of a builtin to compare “at least”.
        * Issue 1:  If you declare version N, can you ship some but not all features from N+1
        * Issue 2:  Major version number
* KG: Discussion around putting a constant with the actual number, and use it or don’t use it.
* KG: Also,  reassure to try to not have a ton of versions. Exercise discretion.  Aim to not increment a major version (compatibility break); strong desire.
* MM: Two issues: Philosophical how does it evolve. 2. Do we want to expose a version number. Related but not the same.
* KG: Let’s talk about scalar version number.
* MM: No opinion on number or function.  If we choose a scalar then can’t make breaking changes.
* KG: Not sure it’s exactly right.
* MM: Code that does a version check against a constant would be wrong if we added another axis.
* MM: Sounds like 1 vote for single scalar value.
* AB: We proposed a (major, minor); it’s well understood model.  Having a single scalar meas _if _ we have breaking change we’d have to change the form.
* RC: How does version number interact with the API side.
* MM: WebGPU the API does not have a distinct version number.  Just modify the API.  There will be a function in the API to expose the WGSL version number. That will match what WGSL will say.  There’s one field of information exposed in two places. As convenience for the author.
* RC: When would someone not check the API’s report of the shader version number.  Is it try and fail loop.
* AB: We expect use with const_assert.  For someone intentionally setting a bar “at least this specific version”, instead of trying to debug compiler errors.  And offline tooling that makes use of that can take advantage.
* RC: How do you use it.
* KG: Don’t have to declare it at all. Can use with const_assert anywhere.
* MM: One issue: one version number or two.  Second issue: Should be exposed as just constants, or have “at least” comparison.
* KG: Think it’s fine to have one version, slightly better to have two.
* AB: We have preference to use a builtin function.
* KG: Don’t want a situation where you try to reverse engineer with a comparison function.
* MM: counterargument is comparing complicated version numbers is nontrivial.
* BC: Why do you think people would want to single out a specific earlier version, or range of versions.
* KG: Because we fail sometimes. I’ve had to do this kind of thing in the past and it was annoying.
* MM: What do you mean “we fail sometimes”.  
* KG: Sometimes we fail at being exactly backward compatibility.  (Deliberately or by accident).  We do this in WebGL. If you don’t have a way to deal with that, then you have to be perfect.
* MM: Scenario:  V1 has no FOO, v2 has FOO but it’s broken, v3 has FOO that is fixed.
* DN: Want to add friction to doing the “bad” thing which is *equality* to a specific version.
* MM: Synthesizing:  **One function that takes one argument (an integer) that returns a bool.**
* KG: Think it’s a bad idea, but can agree.
* JB: To get on the record.  WGSL is not like JS.  JS is late binding, but WGSL is not late binding. You can’t translate the shader if there are any references to a function that is not defined (for example).  Tactics in JS are not going to work in WGSL.  Our plans for having things evolve smoothly may not be sufficient, or needs more mechanism than in JS.  May have to have structural/conditional.
* AB: JS is our preprocessor. We’re exposing the version in the API.  We’re going to do our best to maintain backward compat. Hopefully we can do a good job in forward compatibility. If you need to generate a different shader, you should not encode it in the same string that’s passed to createShaderModule. Hand it the right version of shader.
* MM: We can tell that to authors, as a start.  
* MM: Jim’s stated dis-analogy is worthy to think about.   Will think about it.  True for std library; not true for language features.  JS ‘classes’ for example, is a hard cliff; will cause entire source fail.
* JB: Think using JS to glue source code / munge it is fine. 
* AB: Also want to surface another point from Myle’s feedback on the PR.  One item: all the features that come with a core WGSL version should be exposed in a browser at the same time.  Would be hard to write CTS otherwise.  To allay concerns about delivering features in different bundles, can be addressed with other procedural aspects.
* MM: The 4th piece of feedback that we gave.  Example hypothetical.  Say all browsers have WGSL v3.  WGSL v4 is specified to have 17 new features in it.  WebKit would like to be able to ship some but not all of those 17 features, even if we don’t bump the version number to WGSL v4.  If we implement that feature in v4, don’t want to wait to ship it.  Don’t want to wait to implement unrelated things.
* AB: We don’t know how to write CTS for that.  You write CTS negative test, then shipping some new features will cause that CTS to fail.  If you want to handle that scenario, use an enable for the purpose.  An explicit opt-in “i’m looking for more than what’s necessarily advertised”.  E.g. enable future_stuff_v4;
* MM: Interesting idea.  Thinking off top is to flip the default; explicit opt-in to *not* have the new stuff.  Second idea. Is same basics, but make it an opt-in to not get the additional future features.
* Bc: the point about having a monotonic version is look at the spec and know what the browser will be able to do.  Pulling in future features means that mapping (and number) doesn’t mean much anymore.   Like Alan says, CTS is nearly impossible to write; must handle negative tests.
* JB: Thought the strategy was “we’ll implement stuff”. Premise is that we won’t break things for them.  The reason we have shadowing of builtins is to allow us to grow it.
* AB: Motivating example, we agreed to future feature of better pointers for params.  And CTS has tests that make sure certain advanced uses will fail.  How do you write those CTS tests?
* JB: Myles’ example is introducing new builtins.  Are we going to test future things don’t exist?
* AB: The riskier thing is when you expand an existing feature.
* JB: That’s different from what I recall.  Thought we’d allow implementing new stuff.  Should not have negative tests in CTS.
* KG: Using “can_i_use” is fine. It’s totally workable and fine.
* BC: Following up Jim’s comment. The CTS argument is not that you can’t add new things. CTS can check revisions. 
* AB: the version_at_least is to guarantee something will work because shader only uses features up to that version.
* MM: Hesitant to make forks of CTS for every version number. Procedurally.  Also not really concerned about the CTS.  Care about shipping software to users.  The fact that CTS can spurious failures that happen is not a big deal.  If your shipping criteria is 100% then you’re free to hold back partial versions.
* JB: CTS should not have tests that check for absence of features. It’s not a promise that we make.  That’s not a useful property.  It’s ok to say “please don’t let me use after this point”, and then honor that.
    * Myles agrees
    * Kelsey disagrees
* BC: If your browser says “implement 1.0”, and we have CTS tests that passing uniform and storage buffers is an error.
* KG: This is what I am worried about. I see your line of logic, and you end up making this choice. I worry this is large divergence from WebGL.  WebGL did an incredible job making portale what was not portable before.  I’m concerned about portability picture.  “Works in Chrome”. In WebGL we had much stronger portability.
* AB: I’m surprised by the suggestion there should not be some form of negative tests.  You want the spec says something and we test another.  We’re just going for some random stuff.  Flies in the face of engineering.  If a browser can implement anything at anytime, then version number makes no sense.  Won’t be useful to users.  We thought we’d have a number and you’d have features land together in a bundle. Don’t think the can-i-iuse and try-and-fail model is a good developer experience.  Don’t think we’re going to add negative test for a possible future function.  Say we should have tests that check when the spec says “X doesn’t work” then that X doesn’t work.
* MM: Say WGSL V2 has full pointers.  And we want to write CTS for full pointers.  What would CTS look like. 
* AB: All tests for v1.0 “just work”. V2 tests queries the API saying “i an run v2”, and then try the new functionality.
* MM: What about those negative tests.
* AB: You can conditionalize the tests.
* JB: You can write all the tests you want with the “don’t let me use more than v1”.  When you give it a shader which doesn’t make that declaration, then it should be able to work.
* JB: It’s a little exaggerated to say we’re not doing engineering.  
* AB: When you’re saying the spec says “you can’t do X”, then you should have a test saying “X doesn’t work”.  Don’t want to leave different browsers in the lurch.  That’s where portability problems creep in.   Spec is the agreement of what a particular versio is.
* JB: When you test for the absence of a feature, those tests are the ones that should be self-limiting.
* AB: When you add the new feature, or the break, you have to update the CTS existing tests and new tests.
* MM: Key distinction is whether a feature goes through the full process to define it.  Browsers agree with how it behaves, but not on timeline.
* 


---

</details>
